diff --git a/satellite/src/main/java/com/linbit/linstor/core/devmgr/DeviceHandlerImpl.java b/satellite/src/main/java/com/linbit/linstor/core/devmgr/DeviceHandlerImpl.java
index 49138a8fd..f642d9a58 100644
--- a/satellite/src/main/java/com/linbit/linstor/core/devmgr/DeviceHandlerImpl.java
+++ b/satellite/src/main/java/com/linbit/linstor/core/devmgr/DeviceHandlerImpl.java
@@ -3,6 +3,7 @@ package com.linbit.linstor.core.devmgr;
 import com.linbit.ImplementationError;
 import com.linbit.InvalidNameException;
 import com.linbit.Platform;
+import com.linbit.exceptions.InvalidSizeException;
 import com.linbit.extproc.ExtCmdFactory;
 import com.linbit.linstor.InternalApiConsts;
 import com.linbit.linstor.LinStorException;
@@ -14,6 +15,7 @@ import com.linbit.linstor.api.ApiConsts;
 import com.linbit.linstor.api.SpaceInfo;
 import com.linbit.linstor.api.interfaces.serializer.CtrlStltSerializer;
 import com.linbit.linstor.backupshipping.BackupShippingMgr;
+import com.linbit.linstor.backupshipping.BackupShippingUtils;
 import com.linbit.linstor.clone.CloneService;
 import com.linbit.linstor.core.ControllerPeerConnector;
 import com.linbit.linstor.core.StltExternalFileHandler;
@@ -29,6 +31,7 @@ import com.linbit.linstor.core.objects.Resource;
 import com.linbit.linstor.core.objects.Resource.Flags;
 import com.linbit.linstor.core.objects.ResourceDefinition;
 import com.linbit.linstor.core.objects.Snapshot;
+import com.linbit.linstor.core.objects.SnapshotDefinition;
 import com.linbit.linstor.core.objects.StorPool;
 import com.linbit.linstor.core.objects.Volume;
 import com.linbit.linstor.core.pojos.LocalPropsChangePojo;
@@ -51,6 +54,7 @@ import com.linbit.linstor.netcom.Peer;
 import com.linbit.linstor.propscon.InvalidKeyException;
 import com.linbit.linstor.propscon.InvalidValueException;
 import com.linbit.linstor.propscon.Props;
+import com.linbit.linstor.propscon.ReadOnlyProps;
 import com.linbit.linstor.security.AccessContext;
 import com.linbit.linstor.security.AccessDeniedException;
 import com.linbit.linstor.stateflags.StateFlags;
@@ -83,6 +87,8 @@ import java.util.TreeMap;
 import java.util.TreeSet;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Function;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 
 @Singleton
 public class DeviceHandlerImpl implements DeviceHandler
@@ -299,12 +305,13 @@ public class DeviceHandlerImpl implements DeviceHandler
         return ret;
     }
 
-    private ArrayList<Resource> calculateGrossSizes(Collection<Resource> resources) throws ImplementationError
+    private ArrayList<Resource> calculateGrossSizes(Collection<Resource> resources)
+        throws ImplementationError
     {
         ArrayList<Resource> rscs = new ArrayList<>();
-        try
+        for (Resource rsc : resources)
         {
-            for (Resource rsc : resources)
+            try
             {
                 AbsRscLayerObject<Resource> rscData = rsc.getLayerData(wrkCtx);
                 for (VlmProviderObject<Resource> vlmData : rscData.getVlmLayerObjects().values())
@@ -313,10 +320,19 @@ public class DeviceHandlerImpl implements DeviceHandler
                 }
                 rscs.add(rsc);
             }
-        }
-        catch (AccessDeniedException exc)
-        {
-            throw new ImplementationError(exc);
+            catch (ImplementationError | InvalidSizeException exc)
+            {
+                errorReporter.reportError(
+                    new ImplementationError(
+                        "Caught " + exc.getClass().getSimpleName() + " while processing: " + rsc,
+                        exc
+                    )
+                );
+            }
+            catch (AccessDeniedException exc)
+            {
+                throw new ImplementationError(exc);
+            }
         }
         return rscs;
     }
@@ -493,9 +509,9 @@ public class DeviceHandlerImpl implements DeviceHandler
 
         long rc;
         String errMsg;
-        String cause;
-        String correction;
-        String details;
+        @Nullable String cause;
+        @Nullable String correction;
+        @Nullable String details;
         if (exc instanceof StorageException ||
             exc instanceof ResourceException ||
             exc instanceof VolumeException
@@ -532,7 +548,7 @@ public class DeviceHandlerImpl implements DeviceHandler
             correction = null;
 
             List<String> devLayersAbove = new ArrayList<>();
-            AbsRscLayerObject<?> parent = rscLayerData.getParent();
+            @Nullable AbsRscLayerObject<?> parent = rscLayerData.getParent();
             while (parent != null)
             {
                 devLayersAbove.add(layerFactory.getDeviceLayer(parent.getLayerKind()).getName());
@@ -718,9 +734,10 @@ public class DeviceHandlerImpl implements DeviceHandler
         for (Snapshot snap : snapshots)
         {
             @Nullable ApiCallRcImpl apiCallRc;
+            SnapshotDefinition snapDfn = snap.getSnapshotDefinition();
             try
             {
-                Resource rsc = snap.getResourceDefinition().getResource(wrkCtx, snap.getNodeName());
+                @Nullable Resource rsc = snap.getResourceDefinition().getResource(wrkCtx, snap.getNodeName());
                 apiCallRc = failedRscs.get(rsc);
                 boolean process;
                 if (apiCallRc == null)
@@ -749,7 +766,50 @@ public class DeviceHandlerImpl implements DeviceHandler
                     {
                         // start the backup-shipping-daemons if necessary
                         snapListNotifyApplied.add(snap);
-                        backupShippingManager.allBackupPartsRegistered(snap);
+                        boolean isBackupTarget = BackupShippingUtils.isBackupTarget(snapDfn, snap.getNode(), wrkCtx);
+                        boolean hasShippingStatus = BackupShippingUtils.hasShippingStatus(
+                            snapDfn,
+                            null,
+                            InternalApiConsts.VALUE_SHIPPING,
+                            wrkCtx
+                        );
+                        if (isBackupTarget && hasShippingStatus)
+                        {
+                            @Nullable String srcRemoteName = snapDfn.getSnapDfnProps(wrkCtx)
+                                .getProp(
+                                    InternalApiConsts.KEY_BACKUP_SRC_REMOTE,
+                                    BackupShippingUtils.BACKUP_TARGET_PROPS_NAMESPC
+                                );
+                            if (srcRemoteName != null)
+                            {
+                                backupShippingManager.allBackupPartsRegistered(snap, srcRemoteName);
+                            }
+                        }
+                        else
+                        {
+                            ReadOnlyProps srcProps = snapDfn.getSnapDfnProps(wrkCtx)
+                                .getNamespaceOrEmpty(BackupShippingUtils.BACKUP_SOURCE_PROPS_NAMESPC);
+                            Iterator<String> namespcIter = srcProps.iterateNamespaces();
+                            while (namespcIter.hasNext())
+                            {
+                                String remoteName = namespcIter.next();
+                                @Nullable String dstRemoteName = srcProps.getProp(
+                                    InternalApiConsts.KEY_BACKUP_TARGET_REMOTE,
+                                    remoteName
+                                );
+                                if (
+                                    BackupShippingUtils.hasShippingStatus(
+                                        snapDfn,
+                                        remoteName,
+                                        InternalApiConsts.VALUE_SHIPPING,
+                                        wrkCtx
+                                    )
+                                )
+                                {
+                                    backupShippingManager.allBackupPartsRegistered(snap, dstRemoteName);
+                                }
+                            }
+                        }
                     }
                 }
             }
@@ -804,7 +864,7 @@ public class DeviceHandlerImpl implements DeviceHandler
                 );
             }
             notificationListenerProvider.get()
-                .notifySnapshotDispatchResponse(snap.getSnapshotDefinition().getSnapDfnKey(), apiCallRc);
+                .notifySnapshotDispatchResponse(snapDfn.getSnapDfnKey(), apiCallRc);
         }
     }
 
@@ -847,9 +907,12 @@ public class DeviceHandlerImpl implements DeviceHandler
         {
             try
             {
-                final Props rscDfnProps = rsc.getResourceDefinition().getProps(wrkCtx);
+                final ResourceDefinition rscDfn = rsc.getResourceDefinition();
+                final Props rscDfnProps = rscDfn.getProps(wrkCtx);
                 final @Nullable String srcRscName = rscDfnProps.getProp(InternalApiConsts.KEY_CLONED_FROM);
-                if (srcRscName != null && anyVlmInCloningState(rsc))
+                final boolean isRscDeleting = rsc.getStateFlags().isSet(wrkCtx, Resource.Flags.DELETE) ||
+                    rscDfn.getFlags().isSet(wrkCtx, ResourceDefinition.Flags.DELETE);
+                if (!isRscDeleting && srcRscName != null && anyVlmInCloningState(rsc))
                 {
                     final Resource srcRsc = getResource(rsc, srcRscName);
                     final AbsRscLayerObject<Resource> srcRootLayerObj = srcRsc.getLayerData(wrkCtx);
@@ -866,7 +929,10 @@ public class DeviceHandlerImpl implements DeviceHandler
                 while (vlmsIt.hasNext())
                 {
                     Volume vlm = vlmsIt.next();
-                    if (vlm.getFlags().isSet(wrkCtx, Volume.Flags.CLONING_FINISHED))
+                    StateFlags<Volume.Flags> vlmFlags = vlm.getFlags();
+                    if (isRscDeleting ||
+                        vlmFlags.isSet(wrkCtx, Volume.Flags.DELETE) ||
+                        vlmFlags.isSet(wrkCtx, Volume.Flags.CLONING_FINISHED))
                     {
                         cloneService.removeClone(
                             vlm.getResourceDefinition().getName(),
@@ -1040,7 +1106,7 @@ public class DeviceHandlerImpl implements DeviceHandler
     {
         if (srcStrat.contains(CloneStrategy.ZFS_COPY) && tgtStrat.contains(CloneStrategy.ZFS_COPY))
         {
-            String useZfsClone = targetVlmData.getRscLayerObject().getAbsResource().getResourceDefinition()
+            @Nullable String useZfsClone = targetVlmData.getRscLayerObject().getAbsResource().getResourceDefinition()
                 .getProps(wrkCtx).getProp(InternalApiConsts.KEY_USE_ZFS_CLONE);
             // zfs clone only works if on the same storage pool
             if (StringUtils.propTrueOrYes(useZfsClone) &&
@@ -1081,6 +1147,7 @@ public class DeviceHandlerImpl implements DeviceHandler
             final VlmProviderObject<Resource> targetVlmData = cloneEntry.getValue();
             final Volume targetVlm = (Volume) targetVlmData.getVolume();
             final AbsRscLayerObject<Resource> targetRscData = targetVlmData.getRscLayerObject();
+            final ResourceName resName = targetVlmData.getRscLayerObject().getResourceName();
             try
             {
                 final VolumeNumber vlmNr = targetVlm.getVolumeNumber();
@@ -1115,7 +1182,7 @@ public class DeviceHandlerImpl implements DeviceHandler
 
                     if (cloneStrategy.needsOpenDevices())
                     {
-                        openForClone(sourceVlmData, targetVlmData.getRscLayerObject().getResourceName().displayValue);
+                        openForClone(sourceVlmData, resName.displayValue);
                         openForClone(targetVlmData, null);
                     }
 
@@ -1124,12 +1191,20 @@ public class DeviceHandlerImpl implements DeviceHandler
             }
             catch (StorageException exc)
             {
+                try
+                {
+                    cloneService.cleanupSrcCloneSnapshots(sourceVlmData, resName, false);
+                }
+                catch (StorageException storExc)
+                {
+                    errorReporter.logError("Unable to cleanup source clone snapshots for: ", resName);
+                    errorReporter.reportError(storExc);
+                }
                 cloneService.setFailed(
-                    targetVlmData.getRscLayerObject().getResourceName(),
+                    resName,
                     targetVlmData.getVlmNr(),
                     targetRscData.getResourceNameSuffix());
-                cloneService.notifyCloneStatus(
-                    targetVlmData.getRscLayerObject().getResourceName(), targetVlmData.getVlmNr(), true);
+                cloneService.notifyCloneStatus(resName, targetVlmData.getVlmNr(), true);
                 handleException(targetVlmData.getRscLayerObject().getAbsResource(), exc);
             }
         }
@@ -1527,7 +1602,7 @@ public class DeviceHandlerImpl implements DeviceHandler
                     throw new ImplementationError("Unknown provider kind: " + storPoolInfo.getDeviceProviderKind());
             }
 
-            LocalPropsChangePojo pojo = layer.checkStorPool(storPoolInfo, update);
+            @Nullable LocalPropsChangePojo pojo = layer.checkStorPool(storPoolInfo, update);
             spaceInfo = layer.getStoragePoolSpaceInfo(storPoolInfo);
 
             if (pojo != null)
@@ -1586,7 +1661,7 @@ public class DeviceHandlerImpl implements DeviceHandler
                 .getLayerData(wrkCtx)
                 .getVlmProviderObject(vlmRef.getVolumeNumber());
 
-            TreeSet<String> symlinks = null;
+            @Nullable TreeSet<String> symlinks = null;
             if (!vlmRef.getFlags().isSet(wrkCtx, Volume.Flags.CLONING) &&
                 !vlmRef.getFlags().isSet(wrkCtx, Volume.Flags.DRBD_DELETE))
             {
@@ -1645,8 +1720,11 @@ public class DeviceHandlerImpl implements DeviceHandler
 
     private void updateDiscGran(VlmProviderObject<Resource> vlmData) throws DatabaseException, StorageException
     {
-        String devicePath = vlmData.getDevicePath();
-        if (devicePath != null && vlmData.exists())
+        @Nullable String devicePath = vlmData.getDevicePath();
+        // Check if device path physically exists before calling lsblk
+        // This is important for DRBD devices which might be temporarily unavailable during adjust
+        // (drbdadm adjust brings devices down/up, and kernel might not have created the device node yet)
+        if (devicePath != null && vlmData.exists() && Files.exists(Paths.get(devicePath)))
         {
             if (vlmData.getDiscGran() == VlmProviderObject.UNINITIALIZED_SIZE)
             {
@@ -1713,7 +1791,7 @@ public class DeviceHandlerImpl implements DeviceHandler
         if (isPassthrough != null && isPassthrough)
         {
             final String rscNameSuffix = rscData.getResourceNameSuffix();
-            final AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
+            final @Nullable AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
             final VlmProviderObject<?> childVlmData = childRscData.getVlmProviderObject(vlmData.getVlmNr());
             openForClone(childVlmData, targetRscNameRef);
             vlmData.setCloneDevicePath(childVlmData.getCloneDevicePath());
@@ -1734,7 +1812,7 @@ public class DeviceHandlerImpl implements DeviceHandler
         @Nullable Boolean isPassthrough = rscData.isClonePassthroughMode();
         if (isPassthrough != null && isPassthrough)
         {
-            AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
+            @Nullable AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
             VlmProviderObject<?> childVlmData = childRscData.getVlmProviderObject(vlmData.getVlmNr());
             closeAfterClone(childVlmData, targetRscNameRef);
         }
@@ -1754,7 +1832,7 @@ public class DeviceHandlerImpl implements DeviceHandler
         curLayer.processAfterClone(vlmSrcData, vlmTgtData, clonedDevPath);
 
         final String rscNameSuffix = rscData.getResourceNameSuffix();
-        AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
+        @Nullable AbsRscLayerObject<?> childRscData = rscData.getChildBySuffix(rscNameSuffix);
         if (childRscData != null)
         {
             VlmProviderObject<?> childVlmData = childRscData.getVlmProviderObject(vlmTgtData.getVlmNr());
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/drbd/DrbdLayer.java b/satellite/src/main/java/com/linbit/linstor/layer/drbd/DrbdLayer.java
index a302ee835..0fd255e38 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/drbd/DrbdLayer.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/drbd/DrbdLayer.java
@@ -2,8 +2,8 @@ package com.linbit.linstor.layer.drbd;
 
 import com.linbit.ImplementationError;
 import com.linbit.Platform;
-import com.linbit.PlatformStlt;
 import com.linbit.drbd.DrbdVersion;
+import com.linbit.extproc.ChildProcessHandler;
 import com.linbit.extproc.ExtCmdFactory;
 import com.linbit.extproc.ExtCmdFailedException;
 import com.linbit.linstor.InternalApiConsts;
@@ -12,14 +12,13 @@ import com.linbit.linstor.annotation.Nullable;
 import com.linbit.linstor.api.ApiCallRcImpl;
 import com.linbit.linstor.api.ApiConsts;
 import com.linbit.linstor.api.interfaces.serializer.CtrlStltSerializer;
-import com.linbit.linstor.api.prop.WhitelistProps;
 import com.linbit.linstor.core.ControllerPeerConnector;
-import com.linbit.linstor.core.LinStor;
 import com.linbit.linstor.core.StltConfigAccessor;
 import com.linbit.linstor.core.SysBlockUtils;
 import com.linbit.linstor.core.devmgr.DeviceHandler;
 import com.linbit.linstor.core.devmgr.exceptions.ResourceException;
 import com.linbit.linstor.core.devmgr.exceptions.VolumeException;
+import com.linbit.linstor.core.identifier.ResourceName;
 import com.linbit.linstor.core.identifier.VolumeNumber;
 import com.linbit.linstor.core.objects.AbsResource;
 import com.linbit.linstor.core.objects.Resource;
@@ -35,12 +34,14 @@ import com.linbit.linstor.layer.drbd.drbdstate.DiskState;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdConnection;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdEventPublisher;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdResource;
+import com.linbit.linstor.layer.drbd.drbdstate.DrbdResource.Role;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdStateStore;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdStateTracker;
 import com.linbit.linstor.layer.drbd.drbdstate.DrbdVolume;
 import com.linbit.linstor.layer.drbd.drbdstate.NoInitialStateException;
+import com.linbit.linstor.layer.drbd.drbdstate.ResourceObserver;
 import com.linbit.linstor.layer.drbd.helper.ReadyForPrimaryNotifier;
-import com.linbit.linstor.layer.drbd.utils.ConfFileBuilder;
+import com.linbit.linstor.layer.drbd.resfiles.DrbdResourceFileUtils;
 import com.linbit.linstor.layer.drbd.utils.DrbdAdm;
 import com.linbit.linstor.layer.drbd.utils.MdSuperblockBuffer;
 import com.linbit.linstor.layer.drbd.utils.WindowsFirewall;
@@ -65,18 +66,14 @@ import com.linbit.linstor.storage.utils.MkfsUtils;
 import com.linbit.linstor.storage.utils.VolumeUtils;
 import com.linbit.linstor.utils.layer.DrbdLayerUtils;
 import com.linbit.linstor.utils.layer.LayerRscUtils;
-import com.linbit.utils.AccessUtils;
 
 import javax.inject.Inject;
 import javax.inject.Provider;
 import javax.inject.Singleton;
 
-import java.io.FileOutputStream;
 import java.io.IOException;
 import java.nio.file.Files;
-import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
-import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -85,14 +82,15 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
-import java.util.stream.Collectors;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.event.Level;
 
 @Singleton
 public class DrbdLayer implements DeviceLayer
 {
     public static final String DRBD_DEVICE_PATH_FORMAT = "/dev/drbd%d";
-    private static final String DRBD_CONFIG_SUFFIX = ".res";
-    private static final String DRBD_CONFIG_TMP_SUFFIX = ".res_tmp";
 
     private static final long HAS_VALID_STATE_FOR_PRIMARY_TIMEOUT = 2000;
 
@@ -100,10 +98,10 @@ public class DrbdLayer implements DeviceLayer
 
     private final AccessContext workerCtx;
     private final DrbdAdm drbdUtils;
+    private final DrbdResourceFileUtils drbdResFileUtils;
     private final DrbdStateStore drbdState;
     private final DrbdEventPublisher drbdEventPublisher;
     private final ErrorReporter errorReporter;
-    private final WhitelistProps whitelistProps;
     private final CtrlStltSerializer interComSerializer;
     private final ControllerPeerConnector controllerPeerConnector;
     private final Provider<DeviceHandler> resourceProcessorProvider;
@@ -111,7 +109,6 @@ public class DrbdLayer implements DeviceLayer
     private final StltConfigAccessor stltCfgAccessor;
     private final DrbdVersion drbdVersion;
     private final WindowsFirewall windowsFirewall;
-    private final PlatformStlt platformStlt;
 
     @Nullable private static String drbdSetupStatusOutput;
 
@@ -121,30 +118,30 @@ public class DrbdLayer implements DeviceLayer
      */
     @Nullable private static List<String> adjustResourcesList;
 
+
     @Inject
     public DrbdLayer(
         @DeviceManagerContext AccessContext workerCtxRef,
         DrbdAdm drbdUtilsRef,
+        DrbdResourceFileUtils drbdResFileUtilsRef,
         DrbdStateStore drbdStateRef,
         DrbdEventPublisher drbdEventPublisherRef,
         ErrorReporter errorReporterRef,
-        WhitelistProps whiltelistPropsRef,
         CtrlStltSerializer interComSerializerRef,
         ControllerPeerConnector controllerPeerConnectorRef,
         Provider<DeviceHandler> resourceProcessorRef,
         ExtCmdFactory extCmdFactoryRef,
         StltConfigAccessor stltCfgAccessorRef,
         DrbdVersion drbdVersionRef,
-        WindowsFirewall windowsFirewallRef,
-        PlatformStlt platformStltRef
+        WindowsFirewall windowsFirewallRef
     )
     {
         workerCtx = workerCtxRef;
         drbdUtils = drbdUtilsRef;
+        drbdResFileUtils = drbdResFileUtilsRef;
         drbdState = drbdStateRef;
         drbdEventPublisher = drbdEventPublisherRef;
         errorReporter = errorReporterRef;
-        whitelistProps = whiltelistPropsRef;
         interComSerializer = interComSerializerRef;
         controllerPeerConnector = controllerPeerConnectorRef;
         resourceProcessorProvider = resourceProcessorRef;
@@ -152,7 +149,6 @@ public class DrbdLayer implements DeviceLayer
         stltCfgAccessor = stltCfgAccessorRef;
         drbdVersion = drbdVersionRef;
         windowsFirewall = windowsFirewallRef;
-        platformStlt = platformStltRef;
     }
 
     @Override
@@ -248,7 +244,11 @@ public class DrbdLayer implements DeviceLayer
         DrbdRscData<Resource> drbdRscData = (DrbdRscData<Resource>) rscLayerData;
 
         Resource rsc = rscLayerData.getAbsResource();
-        if (rsc.getProps(workerCtx).map().containsKey(ApiConsts.KEY_RSC_ROLLBACK_TARGET))
+        final Map<String, String> rscPropsMap = rsc.getProps(workerCtx).map();
+
+        condRestartResource(rsc, drbdRscData, rscPropsMap);
+
+        if (rscPropsMap.containsKey(ApiConsts.KEY_RSC_ROLLBACK_TARGET))
         {
             /*
              *  snapshot rollback:
@@ -262,7 +262,7 @@ public class DrbdLayer implements DeviceLayer
                 adjustDrbd(drbdRscData, apiCallRc, true, null);
 
                 // this should not be executed if adjusting the drbd resource fails
-                copyResFileToBackup(drbdRscData);
+                drbdResFileUtils.copyResFileToBackup(drbdRscData);
             }
         }
         else
@@ -274,7 +274,7 @@ public class DrbdLayer implements DeviceLayer
                 processChild(drbdRscData, apiCallRc);
 
                 // this should not be executed if deleting the drbd resource fails
-                deleteBackupResFile(drbdRscData);
+                drbdResFileUtils.deleteBackupResFile(drbdRscData);
             }
             else
             {
@@ -283,7 +283,7 @@ public class DrbdLayer implements DeviceLayer
                     addAdjustedMsg(drbdRscData, apiCallRc);
 
                     // this should not be executed if adjusting the drbd resource fails
-                    copyResFileToBackup(drbdRscData);
+                    drbdResFileUtils.copyResFileToBackup(drbdRscData);
                 }
                 else
                 {
@@ -293,6 +293,136 @@ public class DrbdLayer implements DeviceLayer
         }
     }
 
+    private void condRestartResource(
+        final Resource rsc,
+        final DrbdRscData<Resource> drbdRscData,
+        final Map<String, String> rscPropsMap
+    )
+        throws AccessDeniedException
+    {
+        final ResourceName rscName = rsc.getKey().getResourceName();
+        boolean isDeleting = rsc.getStateFlags()
+            .isSomeSet(workerCtx, Resource.Flags.DELETE, Resource.Flags.DRBD_DELETE);
+
+        if (!isDeleting)
+        {
+            final String restartReqStr = rscPropsMap.getOrDefault(
+                InternalApiConsts.MIN_IO_SIZE_RESTART_DRBD,
+                Boolean.FALSE.toString()
+            );
+            final boolean restartReq = Boolean.parseBoolean(restartReqStr);
+            errorReporter.logDebug(
+                "DrbdLayer processResource: DRBD resource \"%s\", restart request = %b",
+                rscName,
+                restartReq
+            );
+
+            if (restartReq)
+            {
+                @Nullable DrbdResource rscState = null;
+                boolean performRestart = false;
+                try
+                {
+                    rscState = drbdState.getDrbdResource(rscName.displayValue);
+                    if (rscState != null)
+                    {
+                        final boolean hasPrimary = isSomePeerPrimary(rscState);
+                        performRestart = !hasPrimary;
+                        errorReporter.logDebug(
+                            "DrbdLayer processResource: DRBD resource \"%s\", have primary peer = %b",
+                            rscName,
+                            hasPrimary
+                        );
+                    }
+                }
+                catch (NoInitialStateException ignored)
+                {
+                }
+
+                if (performRestart)
+                {
+                    errorReporter.logDebug(
+                        "DrbdLayer processResource: Restarting DRBD resource \"%s\"",
+                        rscName
+                    );
+                    final CountDownLatch syncPoint = new CountDownLatch(1);
+                    final ResourceObserver rscObs = new ResourceObserver()
+                    {
+                        @Override
+                        public void resourceDestroyed(DrbdResource rsc)
+                        {
+                            syncPoint.countDown();
+                        }
+                    };
+
+                    drbdState.addObserver(rscObs, DrbdStateTracker.OBS_RES_DSTR);
+
+                    try
+                    {
+                        // rscState != null is invariant at this point, but probably need the check for spotbugs
+                        // Also, not a bad idea if performRestart behavior changes in the future
+                        if (rscState != null)
+                        {
+                            rscState.suppressDestroyEvent(true);
+                        }
+
+                        drbdUtils.down(drbdRscData);
+                        syncPoint.await(ChildProcessHandler.dfltWaitTimeout, TimeUnit.MILLISECONDS);
+
+                        drbdResFileUtils.regenerateResFile(drbdRscData);
+
+                        drbdUtils.adjust(drbdRscData, false, false, false);
+
+                        errorReporter.logDebug(
+                            "DrbdLayer processResource: DRBD resource \"%s\", reset property, %s = %b",
+                            rscName,
+                            InternalApiConsts.MIN_IO_SIZE_RESTART_DRBD,
+                            false
+                        );
+                        rscPropsMap.put(InternalApiConsts.MIN_IO_SIZE_RESTART_DRBD, Boolean.FALSE.toString());
+                    }
+                    catch (InterruptedException ignored)
+                    {
+                    }
+                    catch (ExtCmdFailedException | StorageException exc)
+                    {
+                        errorReporter.reportError(Level.ERROR, exc);
+                    }
+                    finally
+                    {
+                        if (rscState != null)
+                        {
+                            rscState.suppressDestroyEvent(false);
+                        }
+                        drbdState.removeObserver(rscObs);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks whether a peer (local or remote) is primary
+     *
+     * @return true if a peer (local or remote) is primary
+     */
+    private boolean isSomePeerPrimary(final DrbdResource rscState)
+    {
+        boolean isSomePeerPrimary = rscState.getRole() == Role.PRIMARY;
+        final Map<String, DrbdConnection> connStateMap = rscState.getConnectionsMap();
+        final Iterator<DrbdConnection> iterator = connStateMap.values().iterator();
+        while (!isSomePeerPrimary && iterator.hasNext())
+        {
+            DrbdConnection connState = iterator.next();
+            if (connState.getPeerRole() == Role.PRIMARY)
+            {
+                isSomePeerPrimary = true;
+                break;
+            }
+        }
+        return isSomePeerPrimary;
+    }
+
     private boolean shouldDrbdDeviceBeDeleted(DrbdRscData<Resource> drbdRscData)
         throws AccessDeniedException
     {
@@ -430,10 +560,7 @@ public class DrbdLayer implements DeviceLayer
                 }
                 addDeletedMsg(drbdRscData, apiCallRc);
             }
-            Path resFile = asResourceFile(drbdRscData, false, false);
-            errorReporter.logTrace("Ensuring .res file is deleted: %s ", resFile);
-            Files.deleteIfExists(resFile);
-            drbdRscData.setResFileExists(false);
+            drbdResFileUtils.deleteResFile(drbdRscData);
 
             drbdRscData.setExists(false);
             for (DrbdVlmData<Resource> drbdVlmData : drbdRscData.getVlmLayerObjects().values())
@@ -585,13 +712,34 @@ public class DrbdLayer implements DeviceLayer
             // The .res file might not have been generated in the prepare method since it was
             // missing information from the child-layers. Now that we have processed them, we
             // need to make sure the .res file exists in all circumstances.
-            regenerateResFile(drbdRscData);
+            // However, if the underlying devices are not accessible (e.g., LUKS device is closed
+            // during resource deletion), we skip regenerating the res file to avoid errors
+            boolean canRegenerateResFile = true;
+            if (!skipDisk && !drbdRscData.getAbsResource().isDrbdDiskless(workerCtx))
+            {
+                AbsRscLayerObject<Resource> dataChild = drbdRscData.getChildBySuffix(RscLayerSuffixes.SUFFIX_DATA);
+                if (dataChild != null)
+                {
+                    for (DrbdVlmData<Resource> drbdVlmData : drbdRscData.getVlmLayerObjects().values())
+                    {
+                        VlmProviderObject<Resource> childVlm = dataChild.getVlmProviderObject(drbdVlmData.getVlmNr());
+                        if (childVlm == null || !childVlm.exists() || childVlm.getDevicePath() == null)
+                        {
+                            canRegenerateResFile = false;
+                            break;
+                        }
+                    }
+                }
+            }
+            if (canRegenerateResFile)
+            {
+                drbdResFileUtils.regenerateResFile(drbdRscData);
+            }
 
             // createMetaData needs rendered resFile
             for (DrbdVlmData<Resource> drbdVlmData : createMetaData)
             {
                 createMetaData(drbdVlmData);
-                drbdRscData.setAdjustRequired(true);
             }
 
             try
@@ -759,19 +907,47 @@ public class DrbdLayer implements DeviceLayer
 
                 if (drbdRscData.isAdjustRequired())
                 {
-                    try
+                    // Check if underlying devices are accessible before adjusting
+                    // This is important for encrypted resources (LUKS) where the device
+                    // might be closed during deletion
+                    boolean canAdjust = true;
+                    if (!skipDisk && !drbdRscData.getAbsResource().isDrbdDiskless(workerCtx))
                     {
-                        drbdUtils.adjust(
-                            drbdRscData,
-                            false,
-                            skipDisk,
-                            false
-                        );
+                        AbsRscLayerObject<Resource> dataChild = drbdRscData.getChildBySuffix(RscLayerSuffixes.SUFFIX_DATA);
+                        if (dataChild != null)
+                        {
+                            for (DrbdVlmData<Resource> drbdVlmData : drbdRscData.getVlmLayerObjects().values())
+                            {
+                                VlmProviderObject<Resource> childVlm = dataChild.getVlmProviderObject(drbdVlmData.getVlmNr());
+                                if (childVlm == null || !childVlm.exists() || childVlm.getDevicePath() == null)
+                                {
+                                    canAdjust = false;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    
+                    if (canAdjust)
+                    {
+                        try
+                        {
+                            drbdUtils.adjust(
+                                drbdRscData,
+                                false,
+                                skipDisk,
+                                false
+                            );
+                        }
+                        catch (ExtCmdFailedException extCmdExc)
+                        {
+                            drbdResFileUtils.restoreBackupResFile(drbdRscData);
+                            throw extCmdExc;
+                        }
                     }
-                    catch (ExtCmdFailedException extCmdExc)
+                    else
                     {
-                        restoreBackupResFile(drbdRscData);
-                        throw extCmdExc;
+                        drbdRscData.setAdjustRequired(false);
                     }
                 }
 
@@ -1109,13 +1285,23 @@ public class DrbdLayer implements DeviceLayer
     }
 
     @Override
-    public void resumeIo(AbsRscLayerObject<Resource> rscDataRef) throws ExtCmdFailedException
+    public void resumeIo(AbsRscLayerObject<Resource> rscDataRef, boolean ignoredAsRootLayerRef)
+        throws ExtCmdFailedException
     {
-        drbdUtils.resumeIo((DrbdRscData<Resource>) rscDataRef);
+        var drbdRscDataRef = (DrbdRscData<Resource>) rscDataRef;
+        if (drbdResFileUtils.doesResFileExist(drbdRscDataRef))
+        {
+            drbdUtils.resumeIo(drbdRscDataRef);
+        }
+        else
+        {
+            drbdUtils.resumeIoOnMinorNr(drbdRscDataRef);
+        }
     }
 
     @Override
-    public void suspendIo(AbsRscLayerObject<Resource> rscDataRef) throws ExtCmdFailedException
+    public void suspendIo(AbsRscLayerObject<Resource> rscDataRef, boolean ignoredAsRootLayerRef)
+        throws ExtCmdFailedException
     {
         drbdUtils.suspendIo((DrbdRscData<Resource>) rscDataRef);
     }
@@ -1150,83 +1336,78 @@ public class DrbdLayer implements DeviceLayer
     private boolean hasMetaData(DrbdVlmData<Resource> drbdVlmData)
         throws VolumeException, AccessDeniedException
     {
-        String metaDiskPath = drbdVlmData.getMetaDiskPath();
         boolean externalMd = hasExternalMd(drbdVlmData);
-        if (!externalMd)
-        {
-            // internal meta data
-            metaDiskPath = drbdVlmData.getDataDevice();
-        }
+        String metaDiskPath = externalMd ? drbdVlmData.getMetaDiskPath() : drbdVlmData.getDataDevice();
 
-        MdSuperblockBuffer mdUtils = new MdSuperblockBuffer();
-        try
+        boolean hasMetaData = false;
+        if (metaDiskPath != null)
         {
-            mdUtils.readObject(extCmdFactory, metaDiskPath, externalMd);
-        }
-        catch (IOException exc)
-        {
-            throw new VolumeException(
-                String.format(
-                    "Failed to access DRBD super-block of volume %s/%d",
-                    drbdVlmData.getRscLayerObject().getSuffixedResourceName(),
-                    drbdVlmData.getVlmNr().value
-                ),
-                exc
-            );
-        }
-
-        boolean hasMetaData;
+            MdSuperblockBuffer mdUtils = new MdSuperblockBuffer();
+            try
+            {
+                mdUtils.readObject(extCmdFactory, metaDiskPath, externalMd);
+            }
+            catch (IOException exc)
+            {
+                throw new VolumeException(
+                    String.format(
+                        "Failed to access DRBD super-block of volume %s/%d",
+                        drbdVlmData.getRscLayerObject().getSuffixedResourceName(),
+                        drbdVlmData.getVlmNr().value
+                    ),
+                    exc
+                );
+            }
 
-        if (drbdVlmData.checkMetaData() ||
-            // when adding a disk, DRBD believes that it is diskless but we still need to create metadata
-            !drbdVlmData.hasDisk())
-        {
-            if (mdUtils.hasMetaData())
+            // Assume that meta data is there and do not create meta data if the check is disabled (checkMetaData),
+            // except if there is no disk, because when adding a disk, meta data for a disk must be created
+            // despite the DRBD logic still saying it is diskless
+            hasMetaData = !drbdVlmData.checkMetaData() && drbdVlmData.hasDisk();
+            if (!hasMetaData)
             {
-                boolean isMetaDataCorrupt;
-                try
+                if (mdUtils.hasMetaData())
                 {
-                    isMetaDataCorrupt = !drbdUtils.hasMetaData(
-                        metaDiskPath,
-                        drbdVlmData.getVlmDfnLayerObject().getMinorNr().value,
-                        externalMd ? "flex-external" : "internal"
-                    );
-                }
-                catch (ExtCmdFailedException exc)
-                {
-                    throw new VolumeException(
-                        String.format(
-                            "Failed to check DRBD meta-data integrety of volume %s/%d",
-                            drbdVlmData.getRscLayerObject().getSuffixedResourceName(),
-                            drbdVlmData.getVlmNr().value
-                        ),
-                        exc
-                    );
-                }
-                if (isMetaDataCorrupt)
-                {
-                    throw new VolumeException(
-                        "Corrupted drbd-metadata",
-                        null,
-                        "Linstor has found existing DRBD meta data, " +
-                            "but drbdmeta could not read them",
-                        "Check if the DRBD-utils version match the DRBD kernel version. ",
-                        null
-                    );
-                }
-                else
-                {
-                    hasMetaData = true;
+                    boolean isMetaDataCorrupt;
+                    try
+                    {
+                        isMetaDataCorrupt = !drbdUtils.hasMetaData(
+                            metaDiskPath,
+                            drbdVlmData.getVlmDfnLayerObject().getMinorNr().value,
+                            externalMd ? "flex-external" : "internal"
+                        );
+                    }
+                    catch (ExtCmdFailedException exc)
+                    {
+                        throw new VolumeException(
+                            String.format(
+                                "Failed to check DRBD meta-data integrity of volume %s/%d",
+                                drbdVlmData.getRscLayerObject().getSuffixedResourceName(),
+                                drbdVlmData.getVlmNr().value
+                            ),
+                            exc
+                        );
+                    }
+                    if (isMetaDataCorrupt)
+                    {
+                        throw new VolumeException(
+                            "Corrupted drbd-metadata",
+                            null,
+                            "Linstor has found existing DRBD meta data, " +
+                                "but drbdmeta could not read them",
+                            "Check if the DRBD-utils version matches the DRBD kernel version.",
+                            null
+                        );
+                    }
+                    else
+                    {
+                        hasMetaData = true;
+                    }
                 }
             }
-            else
-            {
-                hasMetaData = false;
-            }
         }
         else
         {
-            hasMetaData = true; // just dont create new meta-data if "checkMetaData" is disabled
+            errorReporter.logDebug("metaDiskPath == null in %s method hasMetaData", getClass().getCanonicalName());
         }
         errorReporter.logTrace("Found metadata: %s", hasMetaData);
         return hasMetaData;
@@ -1246,6 +1427,9 @@ public class DrbdLayer implements DeviceLayer
                 drbdVlmData.getVlmNr().getValue());
             drbdVlmData.setMetaDataIsNew(true);
 
+            DrbdRscData<Resource> drbdRscData = drbdVlmData.getRscLayerObject();
+            drbdRscData.setAdjustRequired(true);
+
             if (DrbdLayerUtils.skipInitSync(workerCtx, drbdVlmData))
             {
                 String currentGi = getCurrentGiFromVlmDfnProp(drbdVlmData);
@@ -1346,13 +1530,13 @@ public class DrbdLayer implements DeviceLayer
                 drbdRscData.setExists(true);
 
                 { // check drbdRole
-                    DrbdResource.Role rscRole = drbdRscState.getRole();
-                    if (rscRole == DrbdResource.Role.UNKNOWN)
+                    Role rscRole = drbdRscState.getRole();
+                    if (rscRole == Role.UNKNOWN)
                     {
                         drbdRscData.setAdjustRequired(true);
                     }
                     else
-                    if (rscRole == DrbdResource.Role.PRIMARY)
+                    if (rscRole == Role.PRIMARY)
                     {
                         drbdRscData.setPrimary(true);
                     }
@@ -1552,31 +1736,28 @@ public class DrbdLayer implements DeviceLayer
         }
     }
 
-    private String readResFile(Path resFilePath) throws IOException
-    {
-        return Files.readString(resFilePath);
-    }
+
 
     private List<String> regenerateAllResFile(Set<AbsRscLayerObject<Resource>> rscDataList)
     {
         List<String> notGeneratedList = new ArrayList<>();
         for (AbsRscLayerObject<Resource> rscLayerObject : rscDataList)
         {
-            DrbdRscData<Resource> drbdRscData = (DrbdRscData<Resource>)rscLayerObject;
+            DrbdRscData<Resource> drbdRscData = (DrbdRscData<Resource>) rscLayerObject;
             try
             {
                 if (drbdRscData.isResFileReady(workerCtx) &&
                     !drbdRscData.getAbsResource().getStateFlags().isSomeSet(
                         workerCtx, Flags.DRBD_DELETE, Flags.DELETE, Flags.INACTIVE))
                 {
-                    regenerateResFile(drbdRscData);
+                    drbdResFileUtils.regenerateResFile(drbdRscData);
                 }
                 else
                 {
                     notGeneratedList.add(drbdRscData.getResourceName().displayValue.toLowerCase());
                 }
             }
-            catch (AccessDeniedException|StorageException exc)
+            catch (AccessDeniedException | StorageException exc)
             {
                 errorReporter.reportError(exc);
                 notGeneratedList.add(drbdRscData.getResourceName().displayValue.toLowerCase());
@@ -1585,208 +1766,6 @@ public class DrbdLayer implements DeviceLayer
         return notGeneratedList;
     }
 
-    /**
-     * Compares the contents of 2 res files if they are equal
-     * Starts comparing, after finding the 'resources "' section, because Linstor generates a header with date.
-     * @param resA String content of first res file to compare
-     * @param resB String content of second res file to compare
-     * @return true if real res file content is the same
-     */
-    private boolean isResFileEqual(String resA, String resB)
-    {
-        boolean equal;
-        int beginA = resA.indexOf("resource \"");
-        int beginB = resB.indexOf("resource \"");
-
-        if (beginA >= 0 && beginB >= 0)
-        {
-            equal = resA.substring(beginA).equals(resB.substring(beginB));
-        }
-        else
-        {
-            throw new ImplementationError("isResFileEqual should only be used for DRBD res files.");
-        }
-        return equal;
-    }
-
-    /**
-     * Writes a new resfile if the content really changed.
-     *
-     * @param drbdRscData
-     * @return True if a new res file was written otherwise false.
-     * @throws AccessDeniedException
-     * @throws StorageException
-     */
-    private boolean regenerateResFile(DrbdRscData<Resource> drbdRscData)
-        throws AccessDeniedException, StorageException
-    {
-        boolean fileWritten = false;
-        Path resFile = asResourceFile(drbdRscData, false, false);
-        Path tmpResFile = asResourceFile(drbdRscData, true, false);
-
-        List<DrbdRscData<Resource>> drbdPeerRscDataList = drbdRscData.getRscDfnLayerObject()
-            .getDrbdRscDataList().stream()
-            .filter(otherRscData -> !otherRscData.equals(drbdRscData) &&
-                AccessUtils.execPrivileged(() -> DrbdLayerUtils.isDrbdResourceExpected(workerCtx, otherRscData)) &&
-                AccessUtils.execPrivileged(
-                    () -> !otherRscData.getAbsResource().getStateFlags().isSet(workerCtx, Resource.Flags.INACTIVE)
-                )
-            )
-            .collect(Collectors.toList());
-
-        String content = new ConfFileBuilder(
-            errorReporter,
-            workerCtx,
-            drbdRscData,
-            drbdPeerRscDataList,
-            whitelistProps,
-            stltCfgAccessor.getReadonlyProps(),
-            drbdVersion
-        ).build();
-
-        String onDiskContent = "resource \"i\"{}";
-        if (drbdRscData.resFileExists())
-        {
-            try
-            {
-                onDiskContent = readResFile(resFile);
-            }
-            catch (NoSuchFileException nsfe)
-            {
-                errorReporter.logWarning("Expected resource file %s did not exist. Rewriting...", resFile.toString());
-                drbdRscData.setResFileExists(false);
-            }
-            catch (IOException exc)
-            {
-                errorReporter.reportError(exc);
-            }
-        }
-
-        if (!isResFileEqual(onDiskContent, content))
-        {
-            try (FileOutputStream resFileOut = new FileOutputStream(tmpResFile.toFile()))
-            {
-                resFileOut.write(content.getBytes());
-            }
-            catch (IOException ioExc)
-            {
-                String ioErrorMsg = ioExc.getMessage();
-                if (ioErrorMsg == null)
-                {
-                    ioErrorMsg = "The runtime environment or operating system did not provide a description of " +
-                        "the I/O error";
-                }
-                throw new StorageException(
-                    "Creation of the DRBD configuration file for resource '" + drbdRscData.getSuffixedResourceName() +
-                        "' failed due to an I/O error",
-                    getAbortMsg(drbdRscData),
-                    "Creation of the DRBD configuration file failed due to an I/O error",
-                    "- Check whether enough free space is available for the creation of the file\n" +
-                        "- Check whether the application has write access to the target directory\n" +
-                        "- Check whether the storage is operating flawlessly",
-                    "The error reported by the runtime environment or operating system is:\n" + ioErrorMsg,
-                    ioExc
-                );
-            }
-
-            try
-            {
-                Files.move(
-                    tmpResFile,
-                    resFile,
-                    StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE
-                );
-                drbdRscData.setResFileExists(true);
-                fileWritten = true;
-            }
-            catch (IOException ioExc)
-            {
-                String ioErrorMsg = ioExc.getMessage();
-                throw new StorageException(
-                    "Unable to move temporary DRBD resource file '" + tmpResFile + "' to resource directory.",
-                    getAbortMsg(drbdRscData),
-                    "Unable to move temporary DRBD resource file due to an I/O error",
-                    "- Check whether enough free space is available for moving the file\n" +
-                        "- Check whether the application has write access to the target directory\n" +
-                        "- Check whether the storage is operating flawlessly",
-                    "The error reported by the runtime environment or operating system is:\n" + ioErrorMsg,
-                    ioExc
-                );
-            }
-            errorReporter.logInfo("DRBD regenerated resource file: %s", resFile);
-        }
-        return fileWritten;
-    }
-
-    private void copyResFile(Path srcPath, Path dstPath, String errMsg, String errCause)
-        throws StorageException
-    {
-        try
-        {
-            Files.copy(srcPath, dstPath, StandardCopyOption.REPLACE_EXISTING);
-        }
-        catch (IOException ioExc)
-        {
-            String ioErrorMsg = ioExc.getMessage();
-            if (ioErrorMsg == null)
-            {
-                ioErrorMsg = "The runtime environment or operating system did not provide a description of " +
-                    "the I/O error";
-            }
-            throw new StorageException(
-                errMsg,
-                errCause,
-                null,
-                "- Check whether enough free space is available for the creation of the file\n" +
-                    "- Check whether the application has write access to the target directory\n" +
-                    "- Check whether the storage is operating flawlessly",
-                "The error reported by the runtime environment or operating system is:\n" + ioErrorMsg,
-                ioExc
-            );
-        }
-    }
-
-    private void copyResFileToBackup(DrbdRscData<Resource> drbdRscData) throws StorageException
-    {
-        Path resFile = asResourceFile(drbdRscData, false, false);
-        Path backupFile = asBackupResourceFile(drbdRscData);
-        String rscName = drbdRscData.getSuffixedResourceName();
-        copyResFile(
-            resFile,
-            backupFile,
-            String.format("Failed to create a backup of the resource file of resource '%s'", rscName),
-            getAbortMsg(drbdRscData)
-        );
-    }
-
-    private void restoreBackupResFile(DrbdRscData<Resource> drbdRscData) throws StorageException
-    {
-        String rscName = drbdRscData.getSuffixedResourceName();
-        errorReporter.logError("Restoring resource file from backup: %s", rscName);
-        Path backupFile = asBackupResourceFile(drbdRscData);
-        Path resFile = asResourceFile(drbdRscData, false, false);
-        copyResFile(
-            backupFile,
-            resFile,
-            String.format("Failed to restore resource file from backup of resource '%s'", rscName),
-            getAbortMsg(drbdRscData)
-        );
-    }
-
-    private void deleteBackupResFile(DrbdRscData<Resource> drbdRscDataRef) throws StorageException
-    {
-        Path resFile = asBackupResourceFile(drbdRscDataRef);
-        errorReporter.logTrace("Deleting res file from backup: %s ", resFile);
-        try
-        {
-            Files.deleteIfExists(resFile);
-        }
-        catch (IOException exc)
-        {
-            throw new StorageException("IOException while removing resource file from backup", exc);
-        }
-    }
-
     private void condInitialOrSkipSync(DrbdRscData<Resource> drbdRscData)
         throws AccessDeniedException, StorageException
     {
@@ -1849,20 +1828,22 @@ public class DrbdLayer implements DeviceLayer
                 }
 
 
-                /*
-                 * since we just created this resource, becoming briefly primary should not be an issue.
-                 * primary needs to be done with --force since we might have configured quorum, but did not give DRBD
-                 * enough time to connect to peers.
-                 *
-                 * we need to be primary even if autoPromote is deactivated to create the filesystem
-                 */
-                try (var ignored = drbdUtils.primaryAutoClose(drbdRscData, true, false))
-                {
-                    MkfsUtils.makeFileSystemOnMarked(errorReporter, extCmdFactory, workerCtx, rsc);
-                }
-                catch (ExtCmdFailedException exc)
+                if (MkfsUtils.needsToCreateFs(rsc, workerCtx))
                 {
-                    throw new StorageException("Failed to become secondary again after creating filesystem", exc);
+                    /*
+                     * primary needs to be done with --force since we might have configured quorum, but did not give
+                     * DRBD enough time to connect to peers.
+                     *
+                     * we need to be primary even if autoPromote is deactivated to create the filesystem
+                     */
+                    try (var ignored = drbdUtils.primaryAutoClose(drbdRscData, true, false))
+                    {
+                        MkfsUtils.makeFileSystemOnMarked(errorReporter, extCmdFactory, workerCtx, rsc);
+                    }
+                    catch (ExtCmdFailedException exc)
+                    {
+                        throw new StorageException("Failed to become secondary again after creating filesystem", exc);
+                    }
                 }
             }
         }
@@ -1920,6 +1901,20 @@ public class DrbdLayer implements DeviceLayer
     {
         waitForValidStateForPrimary(drbdRscData);
 
+        @Nullable DrbdResource drbdResource = null;
+        try
+        {
+            drbdResource = drbdState.getDrbdResource(drbdRscData.getSuffixedResourceName());
+        }
+        catch (NoInitialStateException ignored)
+        {
+        }
+        if (drbdResource != null)
+        {
+            // we need to suppress the next primary event from getting fired to the controller so that the controller
+            // does not freeze the block-size for this primary event.
+            drbdResource.suppressRscEventsWhenPrimary(true);
+        }
         try (var ignored = drbdUtils.primaryAutoClose(drbdRscData, true, false))
         {
             // setting to secondary because of two reasons:
@@ -1942,6 +1937,13 @@ public class DrbdLayer implements DeviceLayer
                 cmdExc
             );
         }
+        finally
+        {
+            if (drbdResource != null)
+            {
+                drbdResource.suppressRscEventsWhenPrimary(false);
+            }
+        }
     }
 
     private void waitForValidStateForPrimary(DrbdRscData<Resource> drbdRscData) throws StorageException
@@ -1994,35 +1996,7 @@ public class DrbdLayer implements DeviceLayer
     /*
      * DELETE method and its utilities
      */
-
-    private Path asResourceFile(DrbdRscData<Resource> drbdRscData, boolean temp, boolean cygwinFormat)
-    {
-        String prefix;
-
-        if (cygwinFormat)
-        {
-            prefix = platformStlt.sysRootCygwin();
-        }
-        else
-        {
-            prefix = platformStlt.sysRoot();
-        }
-
-        return Paths.get(
-            prefix + LinStor.CONFIG_PATH,
-            drbdRscData.getSuffixedResourceName() + (temp ? DRBD_CONFIG_TMP_SUFFIX : DRBD_CONFIG_SUFFIX)
-        );
-    }
-
-    private Path asBackupResourceFile(DrbdRscData<Resource> drbdRscData)
-    {
-        return Paths.get(
-            platformStlt.sysRoot() + LinStor.BACKUP_PATH,
-            drbdRscData.getSuffixedResourceName() + DRBD_CONFIG_SUFFIX
-        );
-    }
-
-    private String getAbortMsg(DrbdRscData<Resource> drbdRscData)
+    public static String getAbortMsg(DrbdRscData<Resource> drbdRscData)
     {
         return "Operations on resource '" + drbdRscData.getSuffixedResourceName() + "' were aborted";
     }
@@ -2180,7 +2154,7 @@ public class DrbdLayer implements DeviceLayer
         {
             throw new StorageException("clone DRBD last known block device file failed.", ioExc);
         }
-        catch (ExtCmdFailedException|AccessDeniedException exc)
+        catch (ExtCmdFailedException | AccessDeniedException exc)
         {
             throw new StorageException("drbdmeta check-resize failed", exc);
         }
